### ISSUE TITLE
[Feature Request] bulkAdd() progress callback

### ISSUE THREAD

**Issue by Garethp** [dfahlander/Dexie.js#708]
Hi there. 

I'm using Dexie to import some quarter million records, and for UX I'd like to have a progress bar (since it can take upwards 30 seconds). To achieve this, I was able to make the following modifications to the `bulkAdd` method locally:

 * Have `bulkAdd` accept a third parameter `individualSuccessHook`
 * Changed `if (creatingHook !== nop) {` to  `if (creatingHook !== nop || ( individualSuccessHook !== null &&  individualSuccessHook !== nop)) {`
 * Changed the `hookCtx` assignment to

```js
if ( individualSuccessHook === nop)  individualSuccessHook = null;
hookCtx = { onerror: null, onsuccess:  individualSuccessHook };
```

With this I've been able to keep the speed of the bulk add but still have a nice progress bar. 

I think this could be a valuable feature to add to the main library, but the fact that it's not yet implemented suggests that perhaps it's not one you want to add. Let me know if you do want it, I'd be happy to make a small PR

**Comment by dfahlander** [comment]
Thanks for the suggestion. Progress is an important usability part when importing large amount of data the first time. However, one of the secrets with the speed of bulkAdd() is that it is not subscribing to onsuccess event at all, except for the last row. That way, the browser thread won't be interrupted. 

There's also another issue with sending too large number of rows to bulkAdd - it occupies lot of RAM. This may not be a problem on a desktop computer, but may fail on a mobile device. It would be better to import data in chunks of maybe 10,000 rows per chunk (calling bulkAdd() for each chunk and awaiting the Promise). Preferably not downloading the entire JSON from the server in one string either, if possible.

**Comment by Garethp** [comment]
> However, one of the secrets with the speed of bulkAdd() is that it is not subscribing to onsuccess event at all, except for the last row. That way, the browser thread won't be interrupted

Yup, I did notice this. To work around this issue, I only send an update to the DOM every 1%. Something like this

```js
let lastPercentage = 0;
const updateProgressBar = progress => {};
const updateHook = (currentCount, totalCount) => {
    const percentage = Math.floor((currentCount / totalAmount) * 100);

    if (percentage > lastPercentage) updateProgressBar(percentage);
    lastPercentage = percentage;
};
```

This method doesn't lock up the browser, and I can't see any noticeable performance implications, even after 250k rows. So perhaps incorporating it as a progress hook that only gets notified every 1% rather than on each row add might be more acceptable?

> There's also another issue with sending too large number of rows to bulkAdd - it occupies lot of RAM. 

I have noticed this, and I've been looking for ways to reduce the stress. Thankfully the import is a run on setup kind of import, and for now it hasn't been an issue (unless I try to import closer to half a million or a million rows). But that's probably a good idea that I'll try anyway

> Preferably not downloading the entire JSON from the server in one string either, if possible.

I wasn't aware that JS could read JSON as a stream. I'll look that up, because that could actually be a really good method to use. Previously I tried to import as a CSV, but splitting each row and mapping it to an object had turned out to be more of a performance issue than just loading the entire JSON object

I do have another issue with importing that large of an amount though, and that's that Dexie actually does lock up the browser just on creating all of those requests. Since the actual `for (var i = 0, l = objects.length; i < l; ++i) {` is synchronous itself, there's a few seconds where the page freezes up before it starts to insert rows. Again, not a large issue since the import is a one-time run, but worth noting all the same

**Comment by dfahlander** [comment]
I see, maybe it could be good to have then. The chunking part would be up to the user in case it is needed. The freezing is not just because of the loop, but because of all the work that indexedDB does natively for each call to IDBObjectStore.add() - it does a lot of work in the same thread that calls it. The best solution for that is to do the bulk import from a WebWorker. But beware posting the data to the WebWorker - it can also be a bottleneck. WebWorker should ideally fetch the data itself from server. If that not possible, try using plain string in the argument to postData because otherwise browser needs to do the structured cloning algorithm to move complex types over - and that takes a lot of CPU.

**Comment by Garethp** [comment]
Thanks for the advice. I might look at it at some point in the future. I'll let you know how I go if I do

**Comment by dfahlander** [comment]
Also notice that just the bare present of an onsuccess callback to an IDBRequest will slow down the browser - even if it's just a function doing nothing. This is due to that the browser needs to get hammered with success events all the time. So if implementing this, we should only ever set req.onsuccess for every 100 request.

**Comment by Garethp** [comment]
So, I played around with this a bit, and here's my findings.

1. I tried streaming with `oboe.js`. This didn't do much good, since the actual inserting into the db didn't occur until after the entire file was downloaded anyway. So even though we were streaming, all the entries were loaded in to memory before they were processed in to the database
2. Holding the entire list of entries in memory didn't actually consume all that much RAM anyway. I'd guess less than a hundred megabytes, but my testing wasn't thorough. To be fair, my objects are small, it's the large amount of them that's the issue
3. It's quicker to do a regular fetch than to try and stream, and that's not where my bottleneck was
4. The memory bottleneck for me existed not in holding all the entries in memory, but in trying to insert them all at once. Taking the approach of bulkAdding them in groups of 5,000 did solve this, and my memory usage went from a multiple gigabytes in use for this operation to 100-200Mb. Unlike straight imports, this usage didn't go up as I tried larger and larger datasets.
5. The downside is that this roughly doubled the amount of time to import my database. A bit of a bummer, but after realizing how much of a memory hog my extension was being there's not much choice

Conclusion: I didn't end up using web workers. Firefox seems to work with this asynchronously enough that the UI isn't blocked, and I get smooth updates, but Chrome is a different story sadly. Streaming didn't seem to be my bottleneck (at my dataset size, your mileage may vary), so I don't bother. Chunking the import is what does all the work for me, but sadly now I lose a lot of the speed I had gained by using the `bulkAdd`. Oh well

**Comment by dfahlander** [comment]
Thanks for sharing your analysis 👍
